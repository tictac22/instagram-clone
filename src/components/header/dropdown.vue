<script setup lang="ts">
import { ref } from "vue"

const expanded = ref(false)

const expandMenu = () => (expanded.value = true)
const collapseMenu = () => (expanded.value = false)
</script>

<template>
	<div v-click-outside="collapseMenu" class="relative ml-6">
		<div class="cursor-pointer" @click="expanded = !expanded">
			<img alt="default" src="/default.jpg" class="w-6 h-6" />
		</div>
		<div
			class="bg-white drop-shadow p-2 w-[192px] rounded absolute -right-[10px] top-[40px] transition-all"
			:class="[
				expanded
					? 'opacity-1 visible pointer-events-auto translate-y-0'
					: 'opacity-0 invisible pointer-events-none -translate-y-3'
			]"
		>
			<div class="triangle -top-[13px] right-[11px]" />
			<p class="px-4 py-2">Profile</p>
			<p class="px-4 py-2">Log Out</p>
		</div>
	</div>
</template>
