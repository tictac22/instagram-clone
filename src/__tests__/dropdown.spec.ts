import Dropdown from "@/components/header/dropdown.vue"
import { describe, expect, it, vi } from "vitest"
import { renderWithPinia } from "./helpers/renderWithPinia"

import { mount } from "@vue/test-utils"
import { globalParams } from "./helpers/dropdown.constants"

vi.mock("vue-router", () => ({
	useRoute: vi.fn(),
	useRouter: vi.fn(() => ({
		push: () => {}
	}))
}))

const enum dataAttrs {
	dataFalse = "[data-test='false']",
	dataTrue = "[data-test='true']",
	clickId = "#click"
}
describe("Dropdown component", () => {
	it("dropdown succesffuly mounts", async () => {
		const wrapper = mount(Dropdown, {
			global: {
				...globalParams({ id: "" }),
				plugins: [renderWithPinia(mockState(""))]
			}
		})
		expect(wrapper.find(dataAttrs.dataFalse).exists()).toBeTruthy()
	})

	it("dropdown succesffuly mounts with equal id", async () => {
		const wrapper = mount(Dropdown, {
			global: {
				...globalParams({ id: "1" }),
				plugins: [renderWithPinia(mockState("1"))]
			}
		})
		expect(wrapper.element).toMatchSnapshot()
	})

	it("renders dropdown", async () => {
		const wrapper = mount(Dropdown, {
			global: {
				...globalParams({ id: "1" }),
				plugins: [renderWithPinia(mockState(""))]
			}
		})
		await wrapper.find(dataAttrs.clickId).trigger("click")
		expect(wrapper.find(dataAttrs.dataTrue).exists()).toBeTruthy()
	})
})

function mockState(uid: string) {
	return {
		stateName: "counter",
		state: {
			user: {
				photoUrl: "",
				uid
			}
		}
	}
}
