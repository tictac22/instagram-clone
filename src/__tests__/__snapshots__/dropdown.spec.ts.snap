// Vitest Snapshot v1

exports[`Dropdown component > dropdown succesffuly mounts with equal id 1`] = `
<div
  class="relative ml-6"
>
  <div
    class="relative flex cursor-pointer items-center justify-center rounded-full"
    id="click"
    tabindex="0"
  >
    <!--v-if-->
    <img
      alt="default"
      class="h-6 w-6 select-none rounded-full"
      src="/default.jpg"
    />
  </div>
  <div
    class="absolute -right-[10px] top-[40px] z-[1] w-[192px] rounded bg-white p-2 drop-shadow transition-all pointer-events-none invisible -translate-y-3 opacity-0"
    data-test="false"
  >
    <div
      class="triangle -top-[13px] right-[11px]"
    />
    <router-link-stub
      class="block cursor-pointer px-4 py-2 transition-colors hover:bg-gray-200"
      tabindex="0"
      to="/"
    />
    <p
      aria-label="log out"
      class="cursor-pointer px-4 py-2 transition-colors hover:bg-gray-200"
      tabindex="0"
    >
       Log Out 
    </p>
  </div>
</div>
`;
